File: solutions/javascript/recursion/balancedBinaryTree.js
const getSubtreeIsBalancedAndHeight = (node) => {
    // Base case
    if (node === null) return { isBalanced: true, subtreeHeight: 0 };

    const { isBalanced: leftSubtreeIsBalanced, subtreeHeight: leftSubtreeHeight } =
        getSubtreeIsBalancedAndHeight(node.left);
    const { isBalanced: rightSubtreeIsBalanced, subtreeHeight: rightSubtreeHeight } =
        getSubtreeIsBalancedAndHeight(node.right);

    const curSubtreeIsBalanced = Math.abs(leftSubtreeHeight - rightSubtreeHeight) <= 1;
    const isBalanced = leftSubtreeIsBalanced && rightSubtreeIsBalanced && curSubtreeIsBalanced;

    const curSubtreeHeight = Math.max(leftSubtreeHeight, rightSubtreeHeight) + 1;

    return {
        isBalanced,
        subtreeHeight: curSubtreeHeight,
    };
};

const isBalanced = (node) => {
    const { isBalanced, subtreeHeight } = getSubtreeIsBalancedAndHeight(node);

    return isBalanced;
};

File: solutions/python/recursion/balanced_binary_tree.py
class Solution:
    def get_subtree_is_balanced_and_height(self, node):
        # Base case
        if node is None:
            return True, 0

        left_subtree_is_balanced, left_subtree_height = self.get_subtree_is_balanced_and_height(
            node.left)
        right_subtree_is_balanced, right_subtree_height = self.get_subtree_is_balanced_and_height(
            node.right)

        cur_subtree_is_balanced = abs(left_subtree_height - right_subtree_height) <= 1
        is_balanced = left_subtree_is_balanced and right_subtree_is_balanced and cur_subtree_is_balanced

        cur_subtree_height = max(left_subtree_height, right_subtree_height) + 1

        return is_balanced, cur_subtree_height

    def isBalanced(self, node: Optional[TreeNode]) -> bool:
        is_balanced, _subtree_height = self.get_subtree_is_balanced_and_height(node)

        return is_balanced